Source/WebCore/rendering/style/RenderStyle.cpp

```diff
@@ -1281,19 +1281,19 @@ bool RenderStyle::changeRequiresRepaint(const RenderStyle& other, OptionSet<Styl
         return true;
 
 
-    if (m_nonInheritedData.ptr() != other.m_nonInheritedData.ptr()) {
-        if (m_nonInheritedData->backgroundData.ptr() != other.m_nonInheritedData->backgroundData.ptr()) {
-            if (!m_nonInheritedData->backgroundData->isEquivalentForPainting(*other.m_nonInheritedData->backgroundData, currentColorDiffers))
-                return true;
-        }
+    bool nonInheritedDataDiffers = m_nonInheritedData.ptr() != other.m_nonInheritedData.ptr();
+    if (nonInheritedDataDiffers || currentColorDiffers) {
+        if (!m_nonInheritedData->backgroundData->isEquivalentForPainting(other.m_nonInheritedData->backgroundData.ptr(), currentColorDiffers))
+            return true;
 
-        if (m_nonInheritedData->surroundData.ptr() != other.m_nonInheritedData->surroundData.ptr()) {
-            if (!m_nonInheritedData->surroundData->border.isEquivalentForPainting(other.m_nonInheritedData->surroundData->border, currentColorDiffers))
-                return true;
-        }
+        if (m_nonInheritedData->surroundData.ptr() != other.m_nonInheritedData->surroundData.ptr())
+            return m_nonInheritedData->surroundData->border != other.m_nonInheritedData->surroundData->border;
+
+        if (m_nonInheritedData->surroundData->border.hasVisibleCurrentColor())
```

Comment:
You also need to check `other.m_nonInheritedData->surroundData->border`, and if you do that isn't this equivalent to the old code?

Reply:
ah right I wrote this code too fast, this should be done under the assumption that `m_nonInheritedData->surroundData->border == other.m_nonInheritedData->surroundData->border;`